File: sdk/ts/src/mcp/dispatch.ts
/**
 * @module mcp/dispatch
 * @package @hop-top/kit
 *
 * Era detection + dispatch for the dual-spec MCP surface, plus the
 * framework-free handler export.
 *
 * Hosting model (ADR 0043 §2): this module exports a
 * transport-agnostic request handler, not a server. Binding it to
 * node:http, hono, express, fastify, or a Worker is the adopter's
 * job — kit does not own the adopter's HTTP stack, and a pure
 * request→response function is directly testable against the wire
 * fixtures with no socket.
 *
 * Detection implements ADR 0042's normative rules literally: the
 * marker set M1-M4, the two deliberate non-markers, and the
 * precedence chain D1-D4. It deliberately does NOT delegate to the
 * v2 SDK's own `classifyInboundRequest`: that classifier treats a
 * modern `MCP-Protocol-Version` header as a routing signal and
 * routes `initialize`-plus-modern-envelope to the modern era, both
 * of which contradict ADR 0042 (and the wire fixtures, which are the
 * parity contract). The SDK is still the source of the protocol
 * constants and error codes — see types.ts.
 */

import {
  MCP_ERR_INTERNAL,
  META_KEY_PROTOCOL_VERSION,
  HEADER_MCP_METHOD,
  HEADER_MCP_NAME,
  RawJSON,
  resolveMcpConfig,
  type McpBridge,
  type McpMountOptions,
  type ResolvedMcpConfig,
} from './types.js';
import {
  headerValues,
  LegacyMcpHandler,
  parseErrorResponse,
  writeError,
} from './legacy.js';
import { ModernMcpHandler } from './modern.js';

// --- normalized request / response --------------------------------------

/**
 * A normalized inbound request: method, headers, and body bytes. A
 * header may repeat, which the modern header checks treat as
 * significant, so values may be an array.
 */
export interface McpHttpRequest {
  /** HTTP method, e.g. "POST". */
  method: string;
  /** Header names are matched case-insensitively. */
  headers?: Record<string, string | string[] | undefined>;
  /** Raw body bytes, exactly as received — never re-encoded. */
  body?: string;
}

/** A normalized response: status, headers, and body bytes. */
export interface McpHttpResponse {
  status: number;
  headers: Record<string, string>;
  body: string;
}

/** Builds a JSON response with the surface's standard content type. */
export function jsonResponse(status: number, body: string): McpHttpResponse {
  return {
    status,
    headers: { 'Content-Type': 'application/json' },
    body,
  };
}

/** The 202 a modern notification receives: empty body, not processed. */
export function acceptedResponse(): McpHttpResponse {
  return { status: 202, headers: {}, body: '' };
}

// --- parsed request -----------------------------------------------------

/**
 * A decoded JSON-RPC request envelope. `id` is kept as a raw JSON
 * fragment so it round-trips byte-for-byte (Go uses json.RawMessage
 * for exactly this reason), and its absence is distinguishable from
 * an explicit null.
 */
export interface JsonRpcRequest {
  jsonrpc?: string;
  id?: RawJSON;
  method: string;
  params?: unknown;
}

/** Which handler a request routes to. */
export type McpEra = 'legacy' | 'modern';

/**
 * Implements ADR 0042's per-request era detection, precedence D1-D4.
 * Never fails — it only classifies an already-parsed request; D1
 * (parse) is the caller's responsibility.
 *
 * Markers:
 *   M1 — Mcp-Method header present
 *   M2 — Mcp-Name header present
 *   M3 — params._meta contains the reserved protocolVersion key
 *        (key presence only; the value is not inspected here)
 *   M4 — method == "server/discover"
 *
 * Deliberate non-markers, both of which are the subtle part and the
 * most likely source of silent divergence:
 *   - Bare params._meta. 2024-11-05 clients legitimately send
 *     _meta.progressToken and OTel traceparent; only the reserved
 *     protocolVersion key signals modern.
 *   - The MCP-Protocol-Version header, at any value. It predates
 *     2026-07-28, so clients that negotiated down to legacy send it
 *     on every subsequent request; treating it as a modern signal
 *     would serve their handshake and then brick the session.
 */
export function detectMcpEra(
  req: McpHttpRequest,
  rpc: JsonRpcRequest,
): McpEra {
  // D2 — initialize is legacy, unconditionally, even when modern
  // markers are present. The spec's dual-era rule says initialize
  // selects legacy semantics; a confused client gets a working legacy
  // handshake, the most recoverable outcome.
  if (rpc.method === 'initialize') return 'legacy';

  // M4 — method == "server/discover".
  if (rpc.method === 'server/discover') return 'modern';

  // M1 / M2 — header presence.
  if (headerValues(req, HEADER_MCP_METHOD).length > 0) return 'modern';
  if (headerValues(req, HEADER_MCP_NAME).length > 0) return 'modern';

  // M3 — params._meta carries the reserved protocolVersion key.
  if (hasModernMetaMarker(rpc.params)) return 'modern';

  // D4 — no markers: legacy. This is the byte-for-byte preservation
  // path: every request a legacy client can send takes today's exact
  // code path.
  return 'legacy';
}

/**
 * M3: whether params._meta carries the reserved protocolVersion key.
 * Only key presence is tested; the value is never inspected at
 * detection time. Malformed or non-object params/_meta are treated as
 * "no marker" — detection never fails, and shape errors are the
 * handlers' business.
 */
export function hasModernMetaMarker(params: unknown): boolean {
  if (params === null || typeof params !== 'object' || Array.isArray(params)) {
    return false;
  }
  const meta = (params as Record<string, unknown>)._meta;
  if (meta === null || typeof meta !== 'object' || Array.isArray(meta)) {
    return false;
  }
  return META_KEY_PROTOCOL_VERSION in (meta as Record<string, unknown>);
}

// --- handler ------------------------------------------------------------

/** A framework-free MCP request handler. */
export type McpHandler = (
  req: McpHttpRequest,
) => Promise<McpHttpResponse>;

/**
 * Builds the MCP handler for a bridge. Mirrors Go's
 * `MountMCP(bridge, router, opts...)`, minus the router: the caller
 * binds the returned function to whatever HTTP stack it uses.
 *
 * Defaults match Go exactly: both eras enabled, path "/mcp", empty
 * origin allowlist, cache hints ttlMs=0 / cacheScope="private".
 * The two mount-time errors are reproduced rather than absorbed —
 * an explicit empty spec-version set, and a negative ttl or
 * unrecognized cache scope, both throw here.
 */
export function createMcpHandler(
  bridge: McpBridge,
  opts: McpMountOptions = {},
): McpHandler {
  const cfg = resolveMcpConfig(opts);
  const dispatcher = new McpDispatcher(bridge, cfg);
  return (req) => dispatcher.handle(req);
}

/** The resolved mount config for a handler, for adopters that need it. */
export function mcpMountConfig(
  opts: McpMountOptions = {},
): ResolvedMcpConfig {
  return resolveMcpConfig(opts);
}

/**
 * Routes each request to exactly one era's handler. Holds no mutable
 * per-request state; both handlers are safe to share concurrently.
 */
export class McpDispatcher {
  private readonly legacy: LegacyMcpHandler;
  private readonly modern: ModernMcpHandler;

  constructor(
    bridge: McpBridge,
    private readonly cfg: ResolvedMcpConfig,
  ) {
    this.legacy = new LegacyMcpHandler(bridge, cfg);
    this.modern = new ModernMcpHandler(bridge, cfg);
  }

  async handle(req: McpHttpRequest): Promise<McpHttpResponse> {
    const verb = (req.method || 'POST').toUpperCase();
    if (verb !== 'POST') {
      // GET and DELETE answer 405 when the modern era is enabled
      // (spec SHOULD for post-session servers). With only the legacy
      // era mounted there is no such route at all, matching Go, where
      // those verbs are simply never registered.
      if (this.cfg.modernEnabled && (verb === 'GET' || verb === 'DELETE')) {
        return methodNotAllowedResponse();
      }
      return methodNotAllowedResponse();
    }

    // D1 — read and parse the body once. An unreadable body is
    // -32603 @ 400; unparseable JSON is -32700 @ 400 — both
    // byte-identical to the legacy responses, regardless of any
    // headers present.
    const body = req.body;
    if (body === undefined) {
      return writeError(
        undefined,
        {
          code: MCP_ERR_INTERNAL,
          message: 'read request body: unexpected EOF',
        },
        400,
      );
    }

    let rpc: JsonRpcRequest;
    try {
      rpc = parseJsonRpcRequest(body);
    } catch {
      return parseErrorResponse(body);
    }

    // Modern-only mounts route everything to the modern handler and
    // handle it per the normal V1-V9 order — no special-casing of
    // initialize anywhere, since D2 exists to hand a confused client
    // to a legacy handler that is not mounted here.
    if (this.cfg.modernEnabled && !this.cfg.legacyEnabled) {
      return this.modern.serve(req, rpc);
    }
    // Legacy-only mounts ignore markers exactly as the pre-dual-spec
    // surface did.
    if (this.cfg.legacyEnabled && !this.cfg.modernEnabled) {
      return this.legacy.serve(req, rpc);
    }

    switch (detectMcpEra(req, rpc)) {
      case 'modern':
        return this.modern.serve(req, rpc);
      default:
        return this.legacy.serve(req, rpc);
    }
  }
}

/** The 405 GET/DELETE answer at the mount path. */
export function methodNotAllowedResponse(): McpHttpResponse {
  return jsonResponse(
    405,
    '{"jsonrpc":"2.0","error":{"code":-32600,"message":"method not allowed"}}\n',
  );
}

/**
 * Parses one JSON-RPC request, preserving the `id` as raw bytes.
 * Throws on unparseable JSON so the caller can render D1's -32700.
 */
export function parseJsonRpcRequest(body: string): JsonRpcRequest {
  const decoded: unknown = JSON.parse(body);
  if (decoded === null || typeof decoded !== 'object' || Array.isArray(decoded)) {
    // Go decodes a non-object body into its request struct and fails,
    // which surfaces as a parse error at the same place.
    throw new SyntaxError('not a JSON-RPC object');
  }
  const obj = decoded as Record<string, unknown>;
  const out: JsonRpcRequest = {
    method: typeof obj.method === 'string' ? obj.method : '',
  };
  if (typeof obj.jsonrpc === 'string') out.jsonrpc = obj.jsonrpc;
  if ('id' in obj) out.id = new RawJSON(rawIdBytes(body, obj.id));
  if ('params' in obj) out.params = obj.params;
  return out;
}

/**
 * Re-serializes the request id. The id shapes JSON-RPC permits
 * (string, number, null) all round-trip identically through
 * JSON.stringify, which is what Go's json.RawMessage echo produces
 * for the same inputs. `_body` is accepted so a future
 * exactness-sensitive change (e.g. preserving a number's original
 * literal spelling) has the source text to work from.
 */
function rawIdBytes(_body: string, id: unknown): string {
  return JSON.stringify(id) ?? 'null';
}

File: sdk/py/hop_top_kit/mcp/dispatch.py
"""Era detection and dispatch for the dual-spec MCP surface.

One mount serves both protocol revisions. This module decides, per
request, which of the two handlers serves it — and exports the
ASGI-compatible callable an adopter binds to their own HTTP server.

Detection rules are normative and are ported literally rather than
re-derived. The marker set ``M(R)`` for a request is:

===  ===========================================================
M1   HTTP header ``Mcp-Method`` present
M2   HTTP header ``Mcp-Name`` present
M3   body ``params._meta`` contains the reserved key
     ``io.modelcontextprotocol/protocolVersion`` — key presence
     only, the value is not inspected at detection time
M4   body ``method == "server/discover"``
===  ===========================================================

Two **deliberate non-markers**, which are the subtle part and the most
likely source of silent divergence:

- **Bare ``params._meta`` is not a marker.** 2024-11-05 clients
  legitimately send ``_meta.progressToken`` and OTel ``traceparent``.
  Only the reserved ``protocolVersion`` key signals the modern era.
- **The ``MCP-Protocol-Version`` header is not a marker.** It predates
  2026-07-28 — it arrived with the 2025-06-18 transport — so a client
  that negotiated *down* to legacy through the handshake sends it on
  every subsequent request. Treating it as a modern signal would serve
  such a client's handshake and then brick its session. Nothing is lost:
  a conforming modern request always carries M1 and M3.

Precedence, first rule that applies winning:

- **D1 — parse.** Read and decode the body once, here. An unreadable
  body is ``-32603`` at 400; unparseable JSON is ``-32700`` at 400 —
  both byte-identical to the legacy responses, whatever headers are
  present.
- **D2 — ``initialize`` is legacy, unconditionally**, even with modern
  markers present. A confused client gets a working handshake, the most
  recoverable outcome; modern clients never send ``initialize``.
- **D3 — any marker routes modern.** Incomplete or contradictory modern
  requests are *not* demoted to legacy: the modern handler rejects them
  with modern errors, which is exactly what a dual-era client keys on to
  avoid falling back incorrectly.
- **D4 — otherwise legacy.** This is the byte-for-byte preservation
  path.
"""

from __future__ import annotations

import json
from typing import Any

from .bridge import Bridge
from .legacy import LegacyHandler
from .modern import ModernHandler, header_confirmation_gate
from .protocol import (
    CACHE_SCOPE_PRIVATE,
    CACHE_SCOPES,
    DEFAULT_PATH,
    DEFAULT_SERVER_NAME,
    DEFAULT_SERVER_VERSION,
    ERR_INVALID_REQUEST,
    ERR_PARSE,
    HEADER_METHOD,
    HEADER_NAME,
    LEGACY_PROTOCOL_VERSION,
    META_PROTOCOL_VERSION,
    MODERN_PROTOCOL_VERSION,
    Headers,
    Request,
    Response,
    RPCRequest,
    parse_rpc,
    write_error,
)

#: The two eras, as :func:`detect_era` reports them.
ERA_LEGACY = "legacy"
ERA_MODERN = "modern"


def detect_era(request: Request, rpc: RPCRequest) -> str:
    """Classify one already-parsed request as legacy or modern.

    Never fails: it only classifies. D1 (parse) is the caller's job, so
    by the time this runs there is always a well-formed envelope to look
    at.
    """
    # D2 — initialize is legacy, unconditionally.
    if rpc.method == "initialize":
        return ERA_LEGACY
    # M4 — server/discover is a marker on its own.
    if rpc.method == "server/discover":
        return ERA_MODERN
    # M1 / M2 — header presence.
    if request.headers.get(HEADER_METHOD):
        return ERA_MODERN
    if request.headers.get(HEADER_NAME):
        return ERA_MODERN
    # M3 — the reserved _meta key. Presence only; the value is not read.
    if has_modern_meta_marker(rpc.params):
        return ERA_MODERN
    # D4 — no markers.
    return ERA_LEGACY


def has_modern_meta_marker(params: Any) -> bool:
    """Report whether ``params._meta`` carries the reserved version key.

    Malformed or non-object ``params``/``_meta`` count as "no marker":
    detection never fails, and surfacing shape errors is the job of
    whichever handler ends up serving the request.
    """
    if not isinstance(params, dict):
        return False
    meta = params.get("_meta")
    if not isinstance(meta, dict):
        return False
    return META_PROTOCOL_VERSION in meta


class MountError(ValueError):
    """A mount-time configuration refusal.

    Misconfiguration fails fast at mount rather than being absorbed into
    surprising per-request behaviour, matching the surface's
    all-or-nothing mount convention.
    """


class McpSurface:
    """The dual-spec MCP surface: a transport-agnostic request handler.

    :meth:`handle` is the whole contract — a pure function from a
    normalised :class:`~hop_top_kit.mcp.protocol.Request` to a
    :class:`~hop_top_kit.mcp.protocol.Response`, with no socket in sight,
    which is what lets the conformance suite drive every fixture case
    without a server. :meth:`__call__` wraps it as an ASGI application
    for adopters who want to mount it under uvicorn, hypercorn, or a
    Starlette route.

    ASGI rather than WSGI is deliberate: the modern era's streaming
    affordances are not expressible in WSGI, and the official Python MCP
    SDK's modern transport is async.
    """

    def __init__(
        self,
        bridge: Bridge,
        *,
        path: str = DEFAULT_PATH,
        server_name: str = DEFAULT_SERVER_NAME,
        server_version: str = DEFAULT_SERVER_VERSION,
        spec_versions: tuple[str, ...] | None = None,
        cache_ttl_ms: int | None = None,
        cache_scope: str | None = None,
        origin_allowlist: tuple[str, ...] = (),
        confirmation_key: bytes | None = None,
        extensions: tuple[Any, ...] = (),
    ) -> None:
        self.path = path
        self._legacy_enabled, self._modern_enabled = _resolve_spec_versions(spec_versions)

        ttl_ms, scope = _resolve_cache_hints(cache_ttl_ms, cache_scope)
        if confirmation_key is not None and not confirmation_key:
            raise MountError("mcp: confirmation_key: empty key")

        self.legacy = LegacyHandler(
            bridge, server_name=server_name, server_version=server_version
        )

        gate = header_confirmation_gate
        if confirmation_key:
            from .modern_confirm import ElicitationConfirmationGate

            gate = ElicitationConfirmationGate(confirmation_key, bridge)

        advertised: dict[str, Any] = {}
        method_handlers: dict[str, Any] = {}
        for extension in extensions:
            advertised[extension.name] = extension.capability()
            method_handlers.update(extension.methods())

        self.modern = ModernHandler(
            bridge,
            server_name=server_name,
            server_version=server_version,
            cache_ttl_ms=ttl_ms,
            cache_scope=scope,
            origin_allowlist=tuple(origin_allowlist),
            confirmation_gate=gate,
            extensions=advertised,
            method_handlers=method_handlers,
        )
        for extension in extensions:
            extension.bind(self.modern)

    # --- request handling -------------------------------------------------

    def handle(self, request: Request) -> Response:
        """Serve one request. The entire public surface of this class."""
        if request.method != "POST":
            if self._modern_enabled:
                # Post-session servers answer the session-era verbs with
                # 405; the POST route is unaffected.
                return _method_not_allowed()
            return _method_not_allowed()

        # D1 — parse once, here.
        try:
            rpc = parse_rpc(request.body)
        except UnicodeDecodeError as exc:
            return write_error(None, ERR_PARSE, f"parse error: {exc}", 400)
        except ValueError as exc:
            return write_error(None, ERR_PARSE, f"parse error: {_go_json_error(request.body, exc)}", 400)

        if self._modern_enabled and not self._legacy_enabled:
            # Modern only: every request runs the normal V1-V9 order with
            # no special-casing of initialize, so a bare legacy handshake
            # fails modern validation rather than being demoted.
            return self.modern.serve(request, rpc)
        if self._legacy_enabled and not self._modern_enabled:
            # Legacy only: markers are ignored exactly as they were
            # before the modern era existed.
            return self.legacy.serve(request, rpc)

        if detect_era(request, rpc) == ERA_MODERN:
            return self.modern.serve(request, rpc)
        return self.legacy.serve(request, rpc)

    # --- ASGI binding ------------------------------------------------------

    async def __call__(self, scope: dict, receive: Any, send: Any) -> None:
        """ASGI application entry point.

        Deliberately does not bind to any particular server: kit does not
        own the adopter's HTTP stack, and a handler that hard-bound to one
        would force that dependency on every consumer.
        """
        if scope.get("type") != "http":
            raise RuntimeError("mcp: McpSurface only handles ASGI http scopes")

        body = b""
        more = True
        while more:
            message = await receive()
            if message["type"] == "http.disconnect":
                return
            body += message.get("body", b"")
            more = message.get("more_body", False)

        headers = Headers(
            [
                (name.decode("latin-1"), value.decode("latin-1"))
                for name, value in scope.get("headers", [])
            ]
        )
        request = Request(
            method=scope.get("method", "POST"),
            path=scope.get("path", self.path),
            headers=headers,
            body=body,
        )
        response = self.handle(request)
        await send(
            {
                "type": "http.response.start",
                "status": response.status,
                "headers": [
                    (name.encode("latin-1"), value.encode("latin-1"))
                    for name, value in response.headers.items()
                ],
            }
        )
        await send({"type": "http.response.body", "body": response.body})


def mount_mcp(bridge: Bridge, **options: Any) -> McpSurface:
    """Build the MCP surface for ``bridge``.

    The option set mirrors Go's ``MountMCP`` exactly — path, server info,
    spec versions, cache hints, origin allowlist, confirmation key — with
    the same defaults: both eras enabled, path ``/mcp``, no origin check,
    and cache hints ``ttlMs = 0`` / ``cacheScope = "private"``.
    """
    return McpSurface(bridge, **options)


# --- mount-time validation ------------------------------------------------


def _resolve_spec_versions(versions: tuple[str, ...] | None) -> tuple[bool, bool]:
    """Resolve the enabled-era set, refusing bad input at mount time.

    ``None`` means the option was not supplied: both eras. An explicitly
    empty tuple is a refusal, not a synonym for the default — the two are
    different intents and conflating them would silently mount a surface
    that serves nothing.
    """
    if versions is None:
        return True, True
    if len(versions) == 0:
        raise MountError("mcp: spec_versions: at least one spec version required")
    legacy = modern = False
    for version in versions:
        if version == LEGACY_PROTOCOL_VERSION:
            legacy = True
        elif version == MODERN_PROTOCOL_VERSION:
            modern = True
        else:
            raise MountError(f"mcp: spec_versions: unrecognized version {version!r}")
    return legacy, modern


def _resolve_cache_hints(ttl_ms: int | None, scope: str | None) -> tuple[int, str]:
    """Validate and default the cache hints."""
    resolved_ttl = 0 if ttl_ms is None else int(ttl_ms)
    if resolved_ttl < 0:
        raise MountError("mcp: cache_ttl_ms: negative ttl")
    resolved_scope = CACHE_SCOPE_PRIVATE if scope is None else scope
    if resolved_scope not in CACHE_SCOPES:
        raise MountError(f"mcp: cache_scope: unknown cache scope {resolved_scope}")
    return resolved_ttl, resolved_scope


def _method_not_allowed() -> Response:
    """The 405 answer to the session-era verbs (GET, DELETE)."""
    return write_error(None, ERR_INVALID_REQUEST, "method not allowed", 405)


def _go_json_error(body: bytes, exc: ValueError) -> str:
    """Render a decode failure the way Go's ``encoding/json`` words it.

    The parse-error *message text* is part of the wire contract: the
    fixtures pin Go's phrasing byte-for-byte, and Python's own
    ``JSONDecodeError`` text shares none of it. Only the shapes the
    fixtures and the spec's own examples exercise are reproduced
    faithfully; anything else falls back to Go's generic wording, which
    is the honest answer for an input neither runtime has pinned.
    """
    text = body.decode("utf-8", errors="replace")
    if not text.strip():
        return "unexpected end of JSON input"
    stripped = text.lstrip()
    if stripped.startswith("{"):
        # Go reports the first byte that cannot begin an object key.
        rest = stripped[1:].lstrip()
        if rest and rest[0] != '"' and rest[0] != "}":
            return (
                f"invalid character {_go_char(rest[0])} "
                "looking for beginning of object key string"
            )
    if stripped:
        return f"invalid character {_go_char(stripped[0])} looking for beginning of value"
    return "unexpected end of JSON input"


def _go_char(char: str) -> str:
    """Quote one character the way Go's JSON errors do (single quotes)."""
    escaped = json.dumps(char)[1:-1].replace("'", "\\'")
    return f"'{escaped}'"
